### PR TITLE
Optimize PyTorch Classifiers and Object Detectors

### LINKED ISSUE #1637: Update argument `channels_first` in `PyTorchObjectDetector` and related estimators
The argument `channels_first` in `PyTorchObjectDetector` is currently unused. We should either remove it or  set the correct value expected by the supported models.

### LINKED ISSUE #2157: Optimize `fit` and `predict` loops for PyTorch Estimators
**Is your feature request related to a problem? Please describe.**
All PyTorch estimators currently index slice numpy arrays in their loops for the `fit` and `predict` methods. Some of these will even put the data on GPU prior to batching which uses an unnecessary amount of GPU VRAM. A more optimized approach is to use the PyTorch dataloader and only move the current batch in the GPU.

**Describe the solution you'd like**
Modify the `fit` and `predict` methods for all PyTorch estimators to use dataloaders and only move the current batch to the GPU

This will not only optimize the 

**Describe alternatives you've considered**
N/A

**Additional context**
The same should be done for all TensorFlowV2 estimators. Some estimators are already using the TensorFlow dataset, but others are still doing numpy index slicing. This can either be done in the same PR or a separate PR.

### LINKED ISSUE #2173: Remove Extraneous Loop for Loss Gradients Calculation in `PyTorchObjectDetector`
**Is your feature request related to a problem? Please describe.**
Due to the issue reported in <https://github.com/pytorch/vision/issues/4153>, the current implementation of the `PyTorchObjectDetector` loops through each input to calculate the gradients. However, this issue has been resolved in `torch>=1.10` and `torchvision>=0.11` which no longer makes this loop necessary.

**Describe the solution you'd like**
The `loss_gradient` method in the `PyTorchObjectDetector` should be simplified to no longer need to loop through each individual input and use the entire batch similar to what is done in the `PyTorchYolo` estimator. Since the estimator already raises an exception when `torch==[1.8, 1.9]` and `torchvision==[0.9, 0.10]` is used, no other version checks are needed since the bug has been fixed in `torch>=1.10` and `torchvision>=0.11`.

This will most likely change the gradient values as the entire batch is being processed at once, so the unit tests will need to be updated as well.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by f4str**
# Description

Use PyTorch datasets and dataloaders to optimize the `fit` and `predict` methods for the following PyTorch estimators: `PyTorchClassifier`, `PyTorchRegressor`, `PyTorchRandomizedSmoothing`, `PyTorchObjectDetector`, `PyTorchYolo`. This not only speeds up the batching process, but also will only move the current batch to the GPU rather than the entire dataset. This will significantly reduce the VRAM usage.

~Optimize the `predict` method for the `TensorFlowV2Classifier` by using a TensorFlow dataset and the `@tf.function` decorator to speed up inference.~

Significant changes to the `PyTorchObjectDetector` estimator:
* The `channels_first` property is now properly used to transpose inputs accordingly.
* Unify the input preprocessing in a similar way that the `PyTorchYolo` estimator does. This simplifies the implementation and reduces redundant code.
* Remove the extraneous loop when computing the loss gradients. The underlying bug has been fixed as of `torch>=1.10` so the entire batch can be processed at once now. Due to batch processing rather an one sample at a time, the loss gradients have now changed. The unit tests have been adjusted accordingly.

Also removed the PyTorch version checking from the `PyTorchYolo` estimator as the bug only applies to the torchvision models and does not affect the external YOLO models.

Fixes #2157 
Fixes #1637 
Fixes #2173 

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Unit tests for PyTorch and TensorFlow estimators work as expected
- [x] Modified unit tests for `PyTorchObjectDetector`
- [x] Modified unit tests for `PyTorchFasterRCNN`

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by f4str** [comment]
This may also potentially resolve #1943 but other issues may still persist.

File: art/estimators/object_detection/utils.py [hunk 1]
```diff
@@ -88,3 +92,50 @@ def convert_pt_to_tf(y: List[Dict[str, np.ndarray]], height: int, width: int) ->
...
+
+    # Convert images into tensor
+    if isinstance(x, np.ndarray):
+        x_tensor = torch.from_numpy(x)
+    else:
+        x_tensor = x
+
+    # Convert labels into tensor
+    if y is not None and isinstance(y, list) and isinstance(y[0]["boxes"], np.ndarray):
+        y_tensor = []
```
Comment (beat-buesser) [inline art/estimators/object_detection/utils.py:122]: Would we be able to remove the "type: ignore" in line 139 if we add more specific typing here in line 122 like
```suggestion
        y_tensor: List[Dict[str, "torch.Tensor"]] = []
```
If yes, should we move this initialization before this if-block?

**Review by beat-buesser — APPROVED** [review]
Hi @f4str Thank you very much for these improvements to the estimators! They will be very useful for many users of ART. I have one suggestion, what do you think?

File: art/estimators/object_detection/utils.py (hunk 1 above)
Comment (f4str) [reply art/estimators/object_detection/utils.py:122]: The `mypy` issue is actually due to the `None` type that `y` can be. The `type: ignore` was actually here before, I've just abstracted this segment of code out (since it is now reused) and made it a function. 

Attempting to make the typing changes to correct this needs to go a couple levels upwards which can makes thing overcomplicated.

**Comment by f4str** [comment]
Hi @beat-buesser thank you for approving the PR. I have reverted the changes to the `TensorFlowV2Classifier` as the `tests/attacks/evasion/test_sign_opt.py` test case runs for over 4 hours with this optimization. Since this is difficult to debug, it is simpler to omit this change for now (since the main focus of this PR is the PyTorch classifiers) and revisit it later.

I've also responded to your comment about the mypy issue.

File: art/estimators/object_detection/utils.py (hunk 1 above)
Comment (f4str) [reply art/estimators/object_detection/utils.py:124]: This should now be resolved

**Comment by beat-buesser** [comment]
@f4str Thank you very much!
